server/protocols/jmap/src/main/java/org/apache/james/jmap/methods/GetMailboxesMethod.java

```diff
@@ -102,17 +101,19 @@ private GetMailboxesResponse getMailboxesResponse(GetMailboxesRequest mailboxesR
         }
     }
 
-    private Stream<MailboxId> retrieveMailboxIds(Optional<ImmutableList<MailboxId>> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
+    private Stream<Optional<Mailbox>> retrieveMailboxes(Optional<ImmutableList<MailboxId>> mailboxIds, MailboxSession mailboxSession) throws MailboxException {
         if (mailboxIds.isPresent()) {
             return mailboxIds.get()
-                .stream();
+                .stream()
+                .map(mailboxId -> mailboxFactory.fromMailboxId(mailboxId, mailboxSession));
         } else {
             List<MailboxMetaData> userMailboxes = mailboxManager.search(
                 MailboxQuery.builder(mailboxSession).privateUserMailboxes().build(),
                 mailboxSession);
             return userMailboxes
                 .stream()
-                .map(MailboxMetaData::getId);
+                .map(MailboxMetaData::getId)
+                .map(mailboxId -> mailboxFactory.fromMailboxId(mailboxId, userMailboxes, mailboxSession));
```

Comment:
it looks like you reached the point where the factory is doing to many things (because method names are hard to find to express that optional second argument).